Samples/Desktop/D3D12Raytracing/src/D3D12RaytracingSakuraScene/Raytracing.hlsl

```diff
@@ -0,0 +1,427 @@
+    uint reflectHint : write(caller) : read(closesthit, caller);
+};
+
+
+// Retrieve hit world position.
+float3 HitWorldPosition()
+{
+    return WorldRayOrigin() + RayTCurrent() * WorldRayDirection();
+}
+
+    
+// Retrieve attribute at a hit position interpolated from vertex attributes using the hit's barycentrics.
+float3 HitAttribute(float3 vertexAttribute[3], BuiltInTriangleIntersectionAttributes attr)
+{
+    return vertexAttribute[0] +
+    attr.barycentrics.x * (vertexAttribute[1] - vertexAttribute[0]) +
+    attr.barycentrics.y * (vertexAttribute[2] - vertexAttribute[0]);
+}
+
+    
+// Retrieve attribute at a hit position interpolated from vertex attributes using the hit's barycentrics (float2 version).
+float2 HitAttribute(float2 vertexAttribute[3], BuiltInTriangleIntersectionAttributes attr)
+{
+    return vertexAttribute[0] +
+    attr.barycentrics.x * (vertexAttribute[1] - vertexAttribute[0]) +
+    attr.barycentrics.y * (vertexAttribute[2] - vertexAttribute[0]);
+}
+
+    
+// Generate a ray in world space for a camera pixel corresponding to an index from the dispatched 2D grid.
+inline void GenerateCameraRay(uint2 index, out float3 origin, out float3 direction)
+{
+    float2 xy = index + 0.5f; // center in the middle of the pixel.
+    float2 screenPos = xy / DispatchRaysDimensions().xy * 2.0 - 1.0;
+
+    // Invert Y for DirectX-style coordinates.
+    screenPos.y = -screenPos.y;
+
+    // Unproject the pixel coordinate into a ray.
+    float4 world = mul(g_sceneCB.projectionToWorld, float4(screenPos, 0, 1));
+
+    world.xyz /= world.w;
+    origin = g_sceneCB.cameraPosition.xyz;
+    direction = normalize(world.xyz - origin);
+}
+    
+
+[shader("raygeneration")]
+void MyRaygenShader()
+{
+    float3 rayDir;
+    float3 origin;
+    
+    // Generate a ray for a camera pixel corresponding to an index from the dispatched 2D grid.
+    GenerateCameraRay(DispatchRaysIndex().xy, origin, rayDir);
+
+    // Trace the ray.
+    // Set the ray's extents.
+    RayDesc ray;
+    ray.Origin = origin;
+    ray.Direction = rayDir;
+    // Set TMin to a non-zero small value to avoid aliasing issues due to floating - point errors.
+    // TMin should be kept small to prevent missing geometry at close contact areas.
+    ray.TMin = 0.001;
+    ray.TMax = 10000.0;
+
+    RayPayload payload =
+    {
+        float4(0, 0, 0, 0),
+        0,
+        0,
+    };
+
+
+    // If toggled 'S', enable SER  
+    if (g_sceneCB.enableSER == 1)
+    {
+        HitObject hit = HitObject::TraceRay(Scene, RAY_FLAG_CULL_BACK_FACING_TRIANGLES, ~0, 0, 1, 0, ray, payload);
+        uint materialID = hit.LoadLocalRootTableConstant(16);
+        bool isFloor = (materialID == 0);
+        bool isTransparentCube = (materialID == 1);
+        uint numHintBits = 1;
+
+        // If material is the floor, sample the texture to decide if it is reflective or not. This mimics subtle water pooling effects in shaded areas.
+        if (isFloor)
```

Comment:
All of the reflectHint calculation only needs to happen if it will be used, which are when sorting by material or sort by both it appears.